### PR TITLE
SI-7570 top-level codegen for toolboxes

### PR DESCRIPTION
**Pull request by xeno-by**
Provides a way to inject top-level classes, traits and modules into
toolbox universes.

Previously that was impossible, because compile and eval both wrap their
arguments into an enclosing method of a synthetic module, which makes it
impossible to later on refer to any definitions from the outside.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym

**Comment by adriaanm** [comment]
ping @retronym

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 1]
```diff
@@ -268,6 +274,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
         }
       }
 
+      def define(tree: ImplDef): Symbol = {
+        val freeTerms = tree.freeTerms
+        if (freeTerms.nonEmpty) throw ToolBoxError(s"reflective toolbox has failed: cannot have free terms in a top-level definition")
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:279]: Should we tel the user what free terms were detected? This should have a test case.

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 2]
```diff
@@ -268,6 +274,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
...
+        verify(tree)
+        wrapInPackageAndCompile(nextWrapperModuleName(), tree)
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:281]: Should we also allow the user to specify the package name?

File: src/compiler/scala/tools/reflect/ToolBox.scala [hunk 3]
```diff
@@ -93,6 +93,14 @@ trait ToolBox[U <: scala.reflect.api.Universe] {
    */
   def compile(tree: u.Tree): () => Any
 
+  /** Defines a top-level class, trait or module in this ToolBox, returning a reference to the defined entity.
+   *
+   *  This method can be used to generate definitions that will later be re-used by subsequent calls to
+   *  `compile`, `define` or `eval`. To refer to the generated definition, use the resulting symbol
+   *  by wrapping it into an `Ident`.
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBox.scala:100]: We should document the fact that the package name is auto-generated.

Isn't `Ident(sym)` deprecated in the public API?

Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBox.scala:100]: Should also document what symbol is returned for a module. (ie module symbol vs module class symbol)

**Comment by xeno-by** [comment]
Will resubmit after ECOOP

**Comment by xeno-by** [comment]
https://github.com/scala/scala/pull/3406

**Pull request closed without merging**